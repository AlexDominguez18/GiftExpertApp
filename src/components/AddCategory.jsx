import { useState } from 'react'
import { FaPlus } from 'react-icons/fa'

const AddCategory = ({ onNewCategory }) => {

  const [ category, setCategory ] = useState('')

  const handleCategoryChange = (e) => {
    setCategory(e.target.value)
  }

  const handleSubmit = (e) => {
    e.preventDefault()
    
    const categoryTrim = category.trim()

    if (categoryTrim.length === 0)
      return
  
    onNewCategory(category)
    setCategory('')
  }

  return (
    <form className={ addCategoryStyles } onSubmit={ handleSubmit }>
      <input 
        type="text"
        className={ addCategoryInputStyles } 
        value={ category }
        onChange={ handleCategoryChange }
      />
      <button 
        className={addCategoryButtonStyles}
        placeholder="Add category..."
      >
        <FaPlus />
        <span className="ml-2">Add category</span>
      </button>
    </form>
  )
}

const addCategoryStyles = `
  basis-1/2
  flex
  flex-col
`

const addCategoryInputStyles = `
  p-2
  m-5
  bg-slate-200
  text-center
  focus:outline-none
  focus:border-b
  border-black
`

const addCategoryButtonStyles = `
  p-2
  mx-5
  px-5
  bg-slate-600
  text-white
  font-bold
  rounded-md
  inline-flex
  items-center
  text-sm
  justify-center
  hover:bg-slate-800
`

export default AddCategory