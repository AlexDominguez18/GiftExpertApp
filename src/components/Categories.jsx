import { useState, useContext } from 'react'
import { AddCategory, CategoryItem } from './'
import { useCategoriesStorage } from "../hooks/useCategoriesStorage"
import { FilterContext } from '../context/FilterContext'

const Categories = () => {
  const { savedCategories, saveCategorie } = useCategoriesStorage()
  const [ categories, setCategories ] = useState(savedCategories)
  const { changeFilter } = useContext(FilterContext);

  const addCategory = (newCategory) => {
    if (categories.includes(newCategory)) return    
    setCategories([...categories, newCategory])
    saveCategorie(newCategory)
    changeFilter(newCategory)
  }

  return (
    <div className={containerStyles}>
      <h2 className={titleStyles}>Categories</h2>
      <div className="flex flex-wrap">
        <ol className={listStyles}>
          {
            categories.length
              ? categories.map((category) => <CategoryItem key={category} category={category} />)
              : <p className="text-center text-slate-700">No categories</p>
          }
        </ol>
        <AddCategory onNewCategory={ addCategory } />
      </div>
    </div>
  )
}

const containerStyles = `
  container
  p-5
  mt-5
  border border-solid
  bg-slate-100
  rounded-lg
  shadow-lg
  lg:mx-auto
`

const titleStyles = `
  mx-auto
  text-center
  text-slate-700
  text-3xl
  font-bold
`

const listStyles = `
  p-3
  mt-3
  list-none
  basis-1/2
`

export { Categories }
