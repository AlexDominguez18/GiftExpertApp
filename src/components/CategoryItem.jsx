
const CategoryItem = ({category}) => {
  return (
    <li className={itemStyles}>
      {category}
    </li>
  )
}

const itemStyles = `
  p-1
  mt-1
  hover:bg-gray-200 duration-200
  hover:text-xl
  cursor-pointer
  border-b-2 border-solid
`

export default CategoryItem