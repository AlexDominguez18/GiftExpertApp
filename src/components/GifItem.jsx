import PropTypes from "prop-types"

const GifItem = ({ id, title, url }) => {
  return (
    <div className={gifItemStyles}>
      <img className="h-52" src={ url } alt={ title } />
      <p className="text-center">{ title }</p>
    </div>
 ) 
}

GifItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired
}

const gifItemStyles = `
  p-2
  m-2
  border border-solid
  bg-slate-100
  rounded-lg
  shadow-lg
  flex
  flex-col
`

export { GifItem }
