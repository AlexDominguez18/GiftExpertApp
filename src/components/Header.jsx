import SearchGift from "./SearchGift"

const titleStyles = `
  text-center
  basis-1/4
  animate-text
  bg-gradient-to-r from-teal-500 via-purple-500 to-orange-500 
  bg-clip-text text-transparent 
  text-2xl
  font-bold
`

const navStyles = `
  bg-slate-800
  p-3 shadow-lg
  flex-wrap
  lg:flex "
`

const Header = ({ title }) => {
  return (
    <nav className={navStyles}>
      <h1 className={titleStyles}>{ title }</h1>
      <SearchGift />
    </nav>
  )
}

export default Header