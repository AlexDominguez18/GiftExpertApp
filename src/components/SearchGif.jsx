import SearchIcon from "./SearchIcon"

const SearchGif = () => {
  return (
    <div className={ searchIconStyles }>
      <SearchIcon />
      <input 
        className={ inputStyles }
        type="text"
        placeholder="Search gif.." 
      />
    </div>
  )
}

const searchIconStyles = `
  basis-3/4
  relative
`

const inputStyles = `
  pl-10
  py-2
  text-slate-100
  bg-slate-700
  w-full
  md:w-1/4
  rounded-full
  caret-slate-100
  focus:outline-none
  focus:w-full
  duration-500
`

export { SearchGif }