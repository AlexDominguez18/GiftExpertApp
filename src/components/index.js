
export * from './AddCategory'
export * from './GifGrid'
export * from './GifItem'
export * from './Categories'
export * from './CategoryItem'
export * from './Header'
export * from './SearchGif'
export * from './SearchIcon'
export * from './Spinner'