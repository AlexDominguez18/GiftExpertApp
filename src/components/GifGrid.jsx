import { useFetchGifs } from '../hooks/useFetchGifs'
import GifItem from './GifItem'
import Spinner from './Spinner'

const GifGrid = ({ filter }) => {
  const { images, loading } = useFetchGifs(filter)

  return (
    <div className={ gridStyles }>
      <div className="text-center">
        <h3 className="text-3xl text-slate-900">{filter} GIF's</h3>
      </div>
      <div className={ gridBodyStyles }>
      {
        loading 
          ? <Spinner /> 
          : images.map( image => <GifItem key={image.id} {... image} />)
      }
      </div>
    </div>
  )
}

const gridStyles = `
  container
  p-5
  mt-5
  border border-solid
  bg-slate-50
  rounded-lg
  shadow-lg
  lg:mx-auto
`

const gridBodyStyles = `
  mt-5
  flex
  flex-wrap
  justify-center
`

export { GifGrid }