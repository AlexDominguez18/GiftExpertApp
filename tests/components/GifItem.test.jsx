import { render, screen } from "@testing-library/react"
import { GifItem } from "../../src/components/GifItem"

describe('GifItem Component Tests', () => {
  const title = "Title"
  const url = "https://example.com/image.jpgw"

  test('Should match the snapshot', () => {
    const { container } = render(<GifItem title={ title } url={ url } />)
    expect( container ).toMatchSnapshot()
  })

  test('Should show image with URL and ALT', () => {
    render(<GifItem title={ title } url={ url } />)
    const { src, alt } = screen.getByRole('img')
    expect(src).toBe(url)
    expect(alt).toBe(title)
  })

  test('Should show the title', () => {
    render(<GifItem title={ title } url={ url } />)
    expect(screen.getByText(title)).toBeTruthy()
  })

})
