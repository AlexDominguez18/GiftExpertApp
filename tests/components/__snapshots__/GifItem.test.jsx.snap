// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifItem Component Tests Should match the snapshot 1`] = `
<div>
  <div
    class="
  p-2
  m-2
  border border-solid
  bg-slate-100
  rounded-lg
  shadow-lg
  flex
  flex-col
"
  >
    <img
      alt="Title"
      class="h-52"
      src="https://example.com/image.jpgw"
    />
    <p
      class="text-center"
    >
      Title
    </p>
  </div>
</div>
`;
